### DISCUSSION TITLE
Interfacing with i2c shield and pigpio

### DISCUSSION THREAD

**Discussion by jmdelahanty** [wehr-lab/autopilot#145] [Q&A]
Hello fellow Autopilots!

I'm learning how to get Autopilot going for getting a setup in my lab that requires me to send the GPIO pins high in a certain order for a small amount of time. Basically it needs to put each pin high for about half a second and turn it low afterwards.  It will cycle through all the pins serially for 3 minutes and, at the end, send all the pins high for one second before the session ends.

I'm super new to pigpio, but I've successfully set pins to outputs and set different ones high and low together, but getting things to cycle properly without Autopilot looks like it will be a big headache and hard to reconfigure.

I'll make a repository today and edit my post here with what I have so far when it's up.

**Comment by sneakers-the-rat** [comment]
Hello jeremy! 

You can do this in a few ways! You can use the [`gpio.Digital_Out`](https://docs.auto-pi-lot.com/en/latest/hardware/gpio.html#autopilot.hardware.gpio.Digital_Out) class, and just use the `.set` method. This would be the simplest to program and give you some flexibility over how you do it.

So for example, if you have a set of 5 pins and want to write a code to all of them every half a second, you could do something like
```python
from autopilot.hardware.gpio import Digital_Out
from time import sleep

pins = [Digital_Out(pin=i) for i in range(5)]

# turn on the first two pins for half a second then turn on the next two...
codes = [
    {'code':[1,1,0,0,0], 'duration':0.5},
    {'code':[0,0,1,1,0], 'duration':0.5}
]

for code in code:
    for pin, value in zip(pins, code['code']):
        pin.set(value)
    sleep(code['duration'])
```

Or any other variation of directly calling the `set` method with whatever representation of timings/values you come up with.

The downside to that is that it'll be a series of python calls to the pigpio daemon, so there will be a bit of jitter between the calls -- shouldn't be more than a few dozen ms, but that can matter. You can also subclass the `gpio.GPIO` class and make a `gpio.Bank` class or somethign similar that wraps the [`set_bank`](http://abyz.me.uk/rpi/pigpio/python.html#set_bank_1) methods to write many values to many pins within the C daemon, which is a much faster way to write to entire banks of pins. Would be happy to help draft that with you and work on a PR :) 

If you aren't using sound output, you can also use the `wave` functions in pigpio, but that might not give you the mutual synchronization you need.

**Reply by jmdelahanty** [reply]
Next week I will be getting back onto this! Lots of other things have come up in the lab that have kept me busy these past couple months.
